elyra/pipeline/kfp/kfp_authentication.py

```diff
@@ -0,0 +1,739 @@
+                if resp.status_code != HTTPStatus.OK:
+                    raise AuthenticationError('Error redirecting to the DEX login page: '
+                                              f'HTTP status code {resp.status_code}.',
+                                              provider=self._type,
+                                              request_history=request_history)
+                # set the login url
+                dex_login_url = resp.url
+
+            ################
+            # Attempt Dex Login
+            ################
+            resp = s.post(
+                dex_login_url,
+                data={"login": username, "password": password},
+                allow_redirects=True
+            )
+            request_history.append((dex_login_url, resp))
+
+            if len(resp.history) == 0:
+                raise AuthenticationError('The LDAP credentials are probably invalid. '
+                                          f'Update runtime configuration \'{runtime_config_name}\' and try again.',
+                                          provider=self._type,
+                                          request_history=request_history)
+
+            # store the session cookies in a "key1=value1; key2=value2" string
+            return "; ".join([f"{c.name}={c.value}" for c in s.cookies])
+
+        # this code should never be reached; raise an error
+        raise AuthenticationError('An implementation problem was detected for LDAP authentication. '
+                                  'Please create an issue.',
+                                  provider=self._type,
+                                  request_history=request_history)
+
+
+class K8sServiceAccountTokenAuthenticator(AbstractAuthenticator):
+    """
+    Authenticator for Service Account Tokens on Kubernetes.
+    """
+
+    _type = SupportedAuthProviders.KUBERNETES_SERVICEACCOUNT_TOKEN
+
+    def authenticate(self,
+                     kf_endpoint: str,
+                     runtime_config_name: str) -> None:
+        """
+        Verify that service account token authentication can be performed.
+        An AuthenticationError is raised if a problem is encountered that
+        would likely prevent the KFP client from authenticating successfully.
+
+        :param kf_endpoint: Kubeflow API endpoint to verify
+        :type kf_endpoint: str
+        :param runtime_config_name: Runtime configuration name where kf_endpoint is specified
+        :type runtime_config_name: str
+        :raises AuthenticationError: A potential issue was detected that will
+        likely cause a KFP client failure.
+        """
+
+        request_history = []
+
+        # Verify the API endpoint
+        resp = requests.get(kf_endpoint, allow_redirects=True)
+        request_history.append((kf_endpoint, resp))
+        if resp.status_code != HTTPStatus.OK:
+            raise AuthenticationError(f'Error detecting whether Kubeflow server at {kf_endpoint} is secured: '
```

Comment:
We should be checking that the redirect history is empty because for `KUBERNETES_SERVICEACCOUNT_TOKEN` to work, the pipelines endpoint must be the internal `ml-pipeline-ui.kubeflow.svc.cluster.local` address (or equivalent), not through the `istio-ingressgateway`.

Reply:
Fixed!